aspnetcore/blazor/fundamentals/dependency-injection.md

```diff
@@ -961,6 +961,111 @@ The app can register transient disposables without throwing an exception. Howeve
+
+`BlazorServiceAccessor.cs`:
+
+```csharp
+internal static class BlazorServiceAccessor
+{
+    private static readonly AsyncLocal<IServiceProvider> s_blazorServices = new();
+
+    public static IServiceProvider Services
+    {
+        get => s_blazorServices.Value ?? throw new InvalidOperationException(
+            "Blazor services are not available in the current context.");
+        set => s_blazorServices.Value = value;
+    }
+}
+```
+
+To set the value of `BlazorServiceAccessor.Services` automatically when an `async` component method is invoked, create a custom base component that re-implements the three primary asynchronous entry points into Razor component code:
+
+* <xref:Microsoft.AspNetCore.Components.IComponent.SetParametersAsync%2A?displayProperty=nameWithType>
+* <xref:Microsoft.AspNetCore.Components.IHandleEvent.HandleEventAsync%2A?displayProperty=nameWithType>
+* <xref:Microsoft.AspNetCore.Components.IHandleAfterRender.OnAfterRenderAsync%2A?displayProperty=nameWithType>
+
+The following class demonstrates the implementation for the base component.
+  
+`CustomComponentBase.cs`:
+
+```csharp
+using Microsoft.AspNetCore.Components;
+
+public class CustomComponentBase : ComponentBase, IHandleEvent, IHandleAfterRender
+{
+    private bool hasCalledOnAfterRender;
+
+    [Inject]
+    private IServiceProvider Services { get; set; } = default!;
+
+    public override Task SetParametersAsync(ParameterView parameters)
+        => InvokeWithBlazorServiceContext(() => base.SetParametersAsync(parameters));
+
+    private Task IHandleEvent.HandleEventAsync(EventCallbackWorkItem callback, object? arg)
+        => InvokeWithBlazorServiceContext(() =>
+        {
+            var task = callback.InvokeAsync(arg);
+            var shouldAwaitTask = task.Status != TaskStatus.RanToCompletion &&
+                task.Status != TaskStatus.Canceled;
+
+            StateHasChanged();
+
+            return shouldAwaitTask ?
+                CallStateHasChangedOnAsyncCompletion(task) :
+                Task.CompletedTask;
+        });
+
+    private Task IHandleAfterRender.OnAfterRenderAsync()
+        => InvokeWithBlazorServiceContext(() =>
+        {
+            var firstRender = !hasCalledOnAfterRender;
+            hasCalledOnAfterRender |= true;
+
+            OnAfterRender(firstRender);
+
+            return OnAfterRenderAsync(firstRender);
+        });
+
+    private async Task CallStateHasChangedOnAsyncCompletion(Task task)
+    {
+        try
+        {
+            await task;
+        }
+        catch
+        {
+            if (task.IsCanceled)
+            {
+                return;
+            }
+
+            throw;
+        }
+
+        StateHasChanged();
+    }
+
+    private async Task InvokeWithBlazorServiceContext(Func<Task> func)
+    {
+        BlazorServiceAccessor.Services = Services;
+        await func();
```

Comment:
Put this inside a `try..finally` and set `.Services` to null inside the finally

Reply:
IntelliSense is telling me that the injected Blazor `IServiceProvider` is non-nullable, so it's not happy with me trying to assign `null` to it.